src/main/java/io/github/moulberry/notenoughupdates/util/ApiUtil.java

```diff
@@ -139,23 +140,24 @@ public Request postData(String contentType, String data) {
 			return this;
 		}
 
-		private CompletableFuture<URL> buildUrl() {
-			CompletableFuture<URL> fut = new CompletableFuture<>();
+		public CompletableFuture<String> requestString() {
+			CompletableFuture<String> result = new CompletableFuture<>();
+			URL url;
 			try {
-				fut.complete(new URIBuilder(baseUrl)
-					.addParameters(queryArguments)
-					.build()
-					.toURL());
-			} catch (URISyntaxException |
-							 MalformedURLException |
-							 NullPointerException e) { // Using CompletableFuture as an exception monad, isn't that exiting?
-				fut.completeExceptionally(e);
+				url = new URIBuilder(baseUrl).addParameters(queryArguments).build().toURL();
+			} catch (URISyntaxException | MalformedURLException | NullPointerException e) {
+				String message = "Error while building url! baseUrl: '" + this.baseUrl + "',  queryArguments: " + queryArguments;
+				result.completeExceptionally(new RuntimeException(message, e));
+				return result;
 			}
-			return fut;
-		}
+			String link = url.toString();
 
-		public CompletableFuture<String> requestString() {
-			return buildUrl().thenApplyAsync(url -> {
+			CompletableFuture<String> cachedResult = HypixelAPICache.INSTANCE.getCacheFor(link);
+			if (cachedResult != null) return cachedResult;
+
+			HypixelAPICache.INSTANCE.addToCache(link, result);
```

Comment:
Having getCache and addToCache not be an atomic operation may lead to race conditions with multiple requests.

Reply:
I don't understand atomic operations enough to argue with that. I thought, since the actual API call gets made in another thread anyway, there's no chance the two cache methods are out of sync.

Reply:
Doesnt matter all that much. should be pretty rare anyway.